adapt/context.py

```diff
@@ -8,18 +8,26 @@ class ContextManagerFrame(object):
     Manages entities and context for a single frame of conversation.
     Provides simple equality querying.
     """
-    def __init__(self, entities=[], metadata={}):
+    def __init__(self, entities=None, metadata=None):
         self.entities = entities
```

Comment:
Why not just initialize self.entities and self.metadata here instead?  I.e.
if entities is None:
    self.entities = []
else:
    self.entities = entities
if metadata is None:
    self.metadata = {}
else:
    self.metadata = metadata

From that point on you know those have been properly constructed and you don't have to do all the checks below on each usage.